locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/BasicLocationNetworkInfoCustomizer.java

```diff
@@ -0,0 +1,472 @@
+            final AttributeSensor<String> sensor = Sensors.newStringSensor("host.address.public." + i++);
+            if (entity.sensors().get(sensor) == null) {
+                entity.sensors().set(sensor, address);
+            }
+        }
+    }
+
+    // --------------------------------------------------------------------------------------
+
+    /**
+     * Returns the hosts and ports that should be considered when determining the address
+     * to use when connecting to the location by assessing the following criteria:
+     * <ol>
+     *     <li>Use the hostAndPortOverride set in options.</li>
+     *     <li>If the machine is connectable, user credentials are given and the machine is provisioned
+     *     in AWS then use {@link JcloudsLocation#getHostnameAws(NodeMetadata, Optional, Supplier, ConfigBag)}.</li>
+     *     <li>If the machine is connectable and pollForFirstReachableAddress is set in options then use all
+     *     {@link #getReachableAddresses reachable} addresses.</li>
+     *     <li>Use the first address that is resolvable with {@link #isAddressResolvable}.</li>
+     *     <li>Use the first address in the node's public then private addresses.</li>
+     * </ol>
+     */
+    protected Iterable<HostAndPort> getManagementCandidates(
+            JcloudsLocation location, NodeMetadata node, ConfigBag config, ManagementAddressResolveOptions options) {
+        final Optional<HostAndPort> hostAndPortOverride = options.getHostAndPortOverride();
+        boolean lookupAwsHostname = Boolean.TRUE.equals(config.get(JcloudsLocation.LOOKUP_AWS_HOSTNAME));
+        String provider = config.get(JcloudsLocation.CLOUD_PROVIDER);
+        if (provider == null) provider = location.getProvider();
+        int defaultPort;
+        if (options.isWindows()) {
+            defaultPort = config.get(WinRmMachineLocation.USE_HTTPS_WINRM) ? 5986 : 5985;
+        } else {
+            defaultPort = node.getLoginPort();
+        }
+
+        // Will normally have come from port forwarding.
+        if (hostAndPortOverride.isPresent()) {
+            // Don't try to resolve it; just use it
+            int port = hostAndPortOverride.get().hasPort()
+                       ? hostAndPortOverride.get().getPort()
+                       : defaultPort;
+            final HostAndPort override = HostAndPort.fromParts(hostAndPortOverride.get().getHostText(), port);
+            LOG.debug("Using host and port override for management candidates of {}: {}", location, override);
+            return ImmutableList.of(override);
+        }
+
+        // Treat AWS as a special case because the DNS fully qualified hostname in AWS is
+        // (normally?!) a good way to refer to the VM from both inside and outside of the region.
+        // TODO This is a bit weird: if the statement below is true then getHostnameAws will find the first
+        // reachable address, which repeats if case after this one.
```

Comment:
So it logins to the machine to find which IP to login to :).
Note that this is different from the next case. It similarly fetches the first reachable IP, but then connects to it and executes a `curl` command, returning the `hostname` value.
Would make more sense to move this after the code which already figured out the correct IPs and use them to fetch the hostname. I'd expect this to be available from EC2 API so in future could improve it to fetch it directly through jclouds.
For amazon it would return the `hostname`. Wonder how it interacts with `PREFER_PRIVATE`. Seems the return value gets coerced to `InetAddress` (which will resolve the hostname from the POV of the Brooklyn instance) and set on the `MachineLocation`.